core/src/main/java/io/grpc/Metadata.java

```diff
@@ -134,15 +135,20 @@ public Integer parseAsciiString(String serialized) {
   private final Map<String, List<MetadataEntry>> store =
       new LinkedHashMap<String, List<MetadataEntry>>();
 
+  /** The number of headers stored by this metadata.  */
+  private int storeCount;
+
   /**
    * Constructor called by the transport layer when it receives binary metadata.
    */
   // TODO(louiscryan): Convert to use ByteString so we can cache transformations
   @Internal
   public Metadata(byte[]... binaryValues) {
-    for (int i = 0; i < binaryValues.length; i++) {
+    checkArgument(binaryValues.length % 2 == 0,
+        "Non multiple of 2 kv pairs: %s", binaryValues.length);
+    for (int i = 0; i < binaryValues.length; i += 2) {
       String name = new String(binaryValues[i], US_ASCII);
-      storeAdd(name, new MetadataEntry(name.endsWith(BINARY_HEADER_SUFFIX), binaryValues[++i]));
+      storeAdd(name, new MetadataEntry(name.endsWith(BINARY_HEADER_SUFFIX), binaryValues[i + 1]));
```

Comment:
is this agreement, or do you actually want the formatting to be;

    i+1

vs 

    i + 1

I believe the latter is more common (correct per the internal style guide?)